skglm/solvers/gram.py

```diff
@@ -0,0 +1,272 @@
+        Target values.
+
+    penalty : instance of Penalty class
+        Penalty used in the model.
+
+    max_epochs : int, optional
+        Maximum number of CD epochs.
+
+    tol : float, optional
+        The tolerance for the optimization.
+
+    w_init : array, shape (n_features,), optional
+        Initial coefficient vector.
+
+    ws_strategy : ('subdiff'|'fixpoint'), optional
+        The score used to compute the stopping criterion.
+
+    verbose : bool or int, optional
+        Amount of verbosity. 0/False is silent.
+
+    Returns
+    -------
+    w : array, shape (n_features,)
+        Coefficients.
+
+    obj_out : array, shape (n_iter,)
+        Objective value at every outer iteration.
+
+    stop_crit : array, shape (n_alphas,)
+        Value of stopping criterion at convergence along the path.
+    """
+    is_sparse = sparse.issparse(X)
+    n_samples = len(y)
+    n_features = len(X.indptr) - 1 if is_sparse else X.shape[1]
+    all_feats = np.arange(n_features)
+    obj_out = []
+    datafit = Quadratic_32() if X.dtype == np.float32 else Quadratic()
+    if is_sparse:
+        datafit.initialize_sparse(X.data, X.indptr, X.indices, y)
+    else:
+        datafit.initialize(X, y)
+    G = X.T @ X  # gram matrix
+    grads = (X.T @ y - G @ w_init) / len(y) if w_init is not None else X.T @ y / len(y)
+    w = w_init.copy() if w_init is not None else np.zeros(n_features, dtype=X.dtype)
+    for epoch in range(max_epochs):
+        if is_sparse:
+            _cd_epoch_gram_sparse(
+                G.data, G.indptr, G.indices, grads, w, datafit, penalty, n_samples,
+                n_features)
+        else:
+            _cd_epoch_gram(G, grads, w, datafit, penalty, n_samples, n_features)
+        if epoch % 50 == 0:
+            Xw = X @ w
+            p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+
+            if is_sparse:
+                grad = construct_grad_sparse(
+                    X.data, X.indptr, X.indices, y, w, Xw, datafit, all_feats)
+            else:
+                grad = construct_grad(X, y, w, Xw, datafit, all_feats)
+            if ws_strategy == "subdiff":
+                opt_ws = penalty.subdiff_distance(w, grad, all_feats)
+            elif ws_strategy == "fixpoint":
+                opt_ws = dist_fix_point(w, grad, datafit, penalty, all_feats)
+
+            stop_crit = np.max(opt_ws)
+            if max(verbose - 1, 0):
+                print(f"Epoch {epoch + 1}, objective {p_obj:.10f}, "
+                      f"stopping crit {stop_crit:.2e}")
+            if stop_crit <= tol:
+                break
+            obj_out.append(p_obj)
+    return w, np.array(obj_out), stop_crit
+
+
+@njit
+def _cd_epoch_gram(G, grads, w, datafit, penalty, n_samples, n_features):
+    """Run an epoch of coordinate descent in place with gradient update using Gram.
+
+    Parameters
+    ----------
+    G : array, shape (n_features, n_features)
+        Gram matrix.
+
+    grads : array, shape (n_features,)
```

Comment:
`grads` is called `grad` in other parts of the code I believe (it's a single gradient, not many of them)